src/Symfony/Component/Mapper/ObjectMapper.php

```diff
@@ -0,0 +1,227 @@
+        }
+
+        if (!\is_object($mapped)) {
+            throw new RuntimeException('Can not map to a non-object.');
+        }
+
+        /**
+         * @var T $mapped the transform on a class should return T but we can't infer types on attributes
+         */
+        $objectMap[$object] = $mapped;
+
+        $arguments = [];
+        $constructor = $toRefl->getConstructor();
+        foreach ($constructor?->getParameters() ?? [] as $parameter) {
+            $parameterName = $parameter->getName();
+            $property = $toRefl->getProperty($parameterName);
+
+            // The mapped class was probably instantiated in a transform we can't write a readonly property
+            if ($property->isReadOnly() && ($property->isInitialized($mapped) && $property->getValue($mapped))) {
+                continue;
+            }
+
+            $arguments[$parameterName] = $parameter->isDefaultValueAvailable() ? $parameter->getDefaultValue() : null;
+        }
+
+        foreach ($refl->getProperties() as $property) {
+            if ($property->isStatic()) {
+                continue;
+            }
+
+            $propertyName = $property->getName();
+
+            $map = null;
+            foreach ($property->getAttributes(Map::class) as $attr) {
+                $map = $attr->newInstance();
+
+                if (($if = $map->if) && ($fn = $this->getCallable($if)) && !$this->call($fn, null, $object)) {
+                    continue 2;
+                }
+
+                break;
+            }
+
+            $mapToProperty = $map?->to ?? $propertyName;
+            if (!$mapToProperty || !$toRefl->hasProperty($mapToProperty)) {
+                continue;
+            }
+
+            $value = $this->propertyAccessor ? $this->propertyAccessor->getValue($object, $propertyName) : $object->{$propertyName};
+            if ($map && $map->transform) {
+                $value = $this->applyTransforms($map, $value, $object);
+            }
+
+            if (
+                \is_object($value)
+                && ($innerMetadata = $this->metadataFactory->create($value))
+                && ($mapTo = $this->getMapTo($innerMetadata, $value, $object))
+                && (\is_string($mapTo->to) && class_exists($mapTo->to))
+            ) {
+                $value = $this->applyTransforms($mapTo, $value, $object);
+
+                if ($value === $object) {
+                    $value = $mapped;
+                } elseif ($objectMap->contains($value)) {
+                    $value = $objectMap[$value];
+                } else {
+                    $value = $this->map($value, $mapTo->to);
+                }
+            }
+
+            if (\array_key_exists($mapToProperty, $arguments)) {
+                $arguments[$mapToProperty] = $value;
+            } else {
+                $this->propertyAccessor ? $this->propertyAccessor->setValue($mapped, $mapToProperty, $value) : ($mapped->{$mapToProperty} = $value);
```

Comment:
there's an issue with this as the signature is `propertyAccessor->setValue(object|array &$objectOrArray, ...)` although the reference is not used for objects... Psalm therefore says that `$mapped` can be an array and we get issues in the return type. Any ideas on how to properly fix that @dunglas or @nicolas-grekas ?

Reply:
Just ignore the Psalm report?